pkg/target/hostMTCP.go

```diff
@@ -0,0 +1,362 @@
+func (h *HostMTCP) UnlockMTCP() {
+	for c := range h.MTCPs {
+		h.MTCPs[c].Unlock()
+	}
+}
+
+// Stream launches the sending to target host.
+// Exits when queue is closed and sending is finished.
+func (h *HostMTCP) Stream(wg *sync.WaitGroup) {
+	if h.conf.TCPOutBufFlushPeriodSec != 0 {
+		go h.flush(time.Second * time.Duration(h.conf.TCPOutBufFlushPeriodSec))
+	}
+	defer func() {
+		wg.Done()
+		close(h.stop)
+	}()
+
+	// send records from the channel to every connection in round robin fashion
+	i := 0
+	for r := range h.Ch {
+		h.tryToSend(r, &h.MTCPs[i])
+		i++
+		if i == h.NumMTCP {
+			i = 0
+		}
+	}
+
+	// this line is only reached when the host channel was closed
+	h.LockMTCP()
+	defer h.UnlockMTCP()
+	h.tryToFlushIfNecessary()
+}
+
+func (h *HostMTCP) tryToSend(r *rec.RecBytes, conn *MultiConnection) {
+	conn.Lock()
+	defer conn.Unlock()
+
+	// retry until successful
+	for {
+		h.ensureConnection(conn)
+		h.keepConnectionFresh(conn)
+
+		err := conn.SetWriteDeadline(time.Now().Add(
+			time.Duration(h.conf.SendTimeoutSec) * time.Second))
+		if err != nil {
+			h.Lg.Warn("error setting write deadline", zap.Error(err))
+		}
+
+		_, err = conn.W.Write(r.Serialize())
+
+		if err == nil {
+			h.outRecs.Inc()
+			h.outRecsTotal.Inc()
+			//h.processingDuration.Observe(time.Since(r.Received).Seconds())
+			conn.LastConnUse = time.Now() // TODO: This is not the last time conn was used. It is used when buffer is flushed.
+			break
+		}
+
+		h.Lg.Warn("error sending value to host. Reconnect and retry..", zap.Error(err))
+		err = conn.Close()
+		if err != nil {
+			// not retrying here, file descriptor may be lost
+			h.Lg.Error("error closing the connection", zap.Error(err))
+		}
+	}
+}
+
+// Flush periodically flushes the buffer and performs a write.
+func (h *HostMTCP) flush(d time.Duration) {
+	t := time.NewTicker(d)
+	defer t.Stop()
+
+	for {
+		select {
+		case <-h.stop:
+			return
+		case <-t.C:
+			h.LockMTCP()
+			h.tryToFlushIfNecessary()
+			h.UnlockMTCP()
+		}
+	}
+}
+
+// Requires h.Conn.Mutex lock.
+func (h *HostMTCP) tryToFlushIfNecessary() {
+	for c := range h.MTCPs {
+		if h.MTCPs[c].W != nil && h.MTCPs[c].W.Buffered() != 0 {
+			if h.MTCPs[c].Conn == nil {
+				h.ensureConnection(&h.MTCPs[c])
+			} else {
+				h.keepConnectionFresh(&h.MTCPs[c])
+			}
+			err := h.MTCPs[c].W.Flush()
+			if err != nil {
+				h.Lg.Error("error while flushing the host buffer", zap.Error(err), zap.String("target_name", h.Name), zap.Uint16("target_port", h.Port))
+				h.MTCPs[c].Conn = nil
+				h.MTCPs[c].W = nil
+			}
+			h.MTCPs[c].LastConnUse = time.Now()
+		}
+	}
+}
+
+// Requires h.Conn.Mutex lock.
+// This function may take a long time.
+func (h *HostMTCP) keepConnectionFresh(conn *MultiConnection) {
+	// 0 value = don't refresh connections
+	if h.conf.TCPOutConnectionRefreshPeriodSec != 0 {
+		if conn != nil && (time.Since(conn.LastConnUse) > time.Second*time.Duration(h.conf.TCPOutConnectionRefreshPeriodSec)) {
```

Comment:
I think we also need to check for conn.Conn != nil.

Reply:
conn here is conn.Conn

Reply:
Calls are conn.Conn calls but conn is the embedding struct. not the embedded interface.

Reply:
Yes, missed that. Why do I need check for conn.Conn != nil here? because of Close(). I'm doing this test in Close() but that's look hideous, indeed :)
Let me fix that.